### DISCUSSION TITLE
Boards have shipped!

### DISCUSSION THREAD

**Discussion by davepl** [PlummersSoftwareLLC/NightDriverStrip#310] [Announcements]
I sent out a batch of boards today to everyone who responded with the BOARD REQUEST email.  I still have a couple left, so if you're actively working on something and somehow missed the list, let me know what you're working on and I'll see if I can spare a board for it!  I hope to order the next batch this week, so supply should open up soon.

Three types of boards have gone out randomly: I personally flashed each one to test it, though, and they should all work equally.  If you got a Red board, that's the production prototype.  If you got a green board, it's a previous rev, but they only difference was they required some hand soldering so we had to make changes.  They should function the same.

Remember that you have to power the matrix.  I've included a power cable with two matrix ends on it, you'll plug in only one to the matrix.  Then cut and strip the single-wire black and red and connect to the PCB as marked + and -.   You can also use side cutters to trim the crimped-on connectors down to half their side, and they'll fit the connector.

The PCB can be misaligned, so make sure it's center vertically in the port.  Make sure the arrows all point up and to the right, and you install it into the PCB on the left side, which is the input.  All the ones I've shipped so far are preinstalled in the matrix.

If you connect the power connector and then plug it into USB, it should work.  You should be able to plug in your remote control receiver and change the effect with the remote.

To flash it, you would build the 'mesmerizer' project.  You'll see two serial ports present when it's plugged in.  If the port is not auto-detected by PlatformIO, set it to the second of the two with the "upload_port=COMx" line in platformio.ini

With that done you should be able to flash the board.  Note that I likely did not flash the SPIFFS partition so you might lack graphics until you do a build and upload of the filesystem data (via the PlatformIO project menu, under mesmerizer).

If you run into any snags, ask below and I'll see what I can do to help!

**Comment by davepl** [comment]
BTW, I also have a 3D file of a 3D-printed stand for it, if anyone is interested.
![2023-06-07_11-31-25](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/3698967/ad96ce13-02fc-4965-ba37-77b963a082be)

**Reply by DiannaLeason** [reply]
Hi Dave,
Not sure if this is the right place for this...How can I get one of your Magic boards?
Email: 
dianna.leason@att.net

**Comment by robertlipe** [comment]
USPS predicts a delivery from your $HOMEland this weekend. Hooray.

My first 2812 strip arrived today. I also have a 64x64 HUB75 on hand and a bucket o' stuff due by the end of the month from Ali, so I look forward to joining you in mastering blinkage in many forms.  ... and casually using "boustrophedon" while shopping for groceries, describing old dot matrix printing, or other tasks. Looks like I can put one of my S3 boards on sockets of a tiny PCB, bring in some external power, and bring out the strip matrix via .100 headers. MY SSD1306'S OR ST735 OLED/LCD'S should provide faster/lower-overhead debugging or real-time displays about the displays than a plain ole serial port, so I can help with that code if it's not already there.

I have started some work on the network GIF displayer (and, probably eventually, network renderer) so I have starter projects in mind. I've also spent a ridiculous amount of time studying nerd geometry for the Hexagon display mentioned. I think we can make that awesome.

If the HUB75's are too hard for the S3's to drive, I'm not opposed to putting something like CH32V003's as I2C targets for them.  They don't have much memory, but any of the BluePill-class 103's can buffer half a frame of a 64x64 display from an DMA'ed SPI/I2C burst. There are _lots_ of possibly fun spaces to explore. 

I'm open to mentorship roles if someone has time to spend and "just" needs help spelling it in C/C++ or debugging simple hardware, etc.

Looking forward to this!

**Comment by Thehaus64** [comment]
Hi Dave, I just came across you LED video series i followed it line by line and got everything working except the fire effect. I'm debugging it currently.
I would like to build a umbrella like yours for my mother. If you still have an extra board I would love to use it for the umbrella and eventually expand it onto her patio. Thanks for all you hard work making all your videos they are Entertaining and educational!

Thanks Jared

**Reply by robertlipe** [reply]
See if https://github.com/PlummersSoftwareLLC/NightDriverStrip/pull/330 helps. It should help with most builds that require multiple controller pins. 

Sorry that I was thinking Atom Light instead of Umbrella, but the root issues may be the same if you hve multiple strips that aren't acting independently, but rather as one strip that's getting jumbled signals as successive writes clobber the controllers.

**Comment by davepl** [comment]
The mesermizer board only works with the HUB75 matrix, so you wouldn’t want one for the umbrella!  I’d recommend the TFT Feather S3 which has an ESP32 and a nice display and some extra RAM.  It’d be perfect for the umbrella.

- Dave

**Comment by robertlipe** [comment]
I think that Atomic Lamp needs a fix that @stretch911 developed. The CL is
sitting in the review queue right now, but Stretch reported it fixing his
Atomic light.

Which made me slightly happy because I'd offered the day before to fix his
Atomic light. :-)

BTW, Dave, what connector does Mesmerizer uses? It seems to be a pitch
that's different than anything in my collection. I have .100's literally by
the hundreds, but don't recognize this one.

**Comment by davepl** [comment]
Do you mean on the back of the Mesmerizer board?  I think it’s standard 2.54mm pitch isn’t it?  Or where are you referring to?   The M5StickC has a weird one, but I think I have a link for correct ones somewhere..

**Reply by robertlipe** [reply]
"Back" is in the eye of the beholder I suppose. I inventory HUB75 cables, displays, etc. so I have those. 

It's this OTHER connector that's used for JTAG and that bring out the other GPIO lines so I could dangle strips, etc. from it that I don't recognize. Here, take a peek over my shoulder:

[MVIMG_20230624_221915](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/8724975/8d4d7483-6cfd-4832-9992-4417b99c77ee)

**Reply by robertlipe** [reply]
After saying that I'd never seen anything with that pin pitch, I'll be darned if today's mail didn't bring me one. (No, I didn't really think you commissioned your own connector. I've just never seen one..at least not in this application.)

That pitch is the same one used by the Waveshare 2.13" E-ink display, which conveniently includes a cable that goes from 8 of those pins out to dupont-style post connectors. So it's probably enough to attach JTAG (since this isn't the ESP32 that includes JTAG on-chip), but far from enough to capture all the pins brought on on mesmerizer.

Now that I have something more Google-able, I think thats called a "JST 2.0" connector. Makes sense as it implies that's a 2.0mm pitch instead of a 2.54mm pitch. That key tidbit is probably enough to order a variety pack and make my own cable, but 28 pins doesn't seem to be common. I'm hoping you can tell me that's a commodity part for some board out there with an existing variety of pinned out cables/headers/accessories already broken out.

What do you connect to this header? What inspired it?

**Comment by rbergen** [comment]
I think some documentation of the pinout of the 2.0mm pin header might indeed be useful. In part, the answer to what you connect to it is in the picture: some pin pairs are intended to "normally" be bridged by jumpers. 

As to why the 2.0mm pitch was chosen, I think of the participants in this discussion only @davepl knows - I know I don't. I wouldn't be surprised if it was as simple as space. With 13 pin spacings between a double row of 28 pins, this header saves 7 mm in overall length. On a small PCB that is already challenging to route (as Dave mentioned in a few of his videos) that matters.

**Comment by robertlipe** [comment]
Yes, please. It does look like the last few pins of that that includes the
jumpers and probably some power decoupling, is effectively a different
connector so maybe that could have been chunked off. But I assume it's cast
in stone by now and would just like to work with what's there instead of
second guessing it.

Tackling the Pulsar bug would be easier if I (or anyone else) could jack in
any of our zillion JTAG pods, but that header is a barrier. Just as a cold
guess, the lessons learned in (
https://github.com/PlummersSoftwareLLC/NightDriverStrip/discussions/331#discussioncomment-6264050)
point to out-of-bounds access to the LEDs array. That would be easier with
the luxury of an MMU, but we should still be able to get a watchpoint on it
somewhere.

Probably Tuesday or so I'll start building up a board that contains the S3
(so I get JTAG and console "for free" when using the better SoC - the
real-world cost of adding JTAG and adding serial pretty much eclipses the
change in the different SoC module) just so I can get Strip (or a bunch)  +
HUB75 (or at least two, knowing I have cables to chain ten HUB75's which
should probably paralyze the ESP32...) on the same board and building the
i2s/audio In to be compatible with mesmerizer and adding external power
would be convenient for helping with the Hexagon board from Kane. That was
my plan if I couldn't penetrate the strip GPIO pins on Mesmerizer. and fill
the audio tubes from something. Plus, I'm growing increasingly crushed by
vendors selling "2812" strips that are actually 2811's and need 12V to feed
them ... after I stocked up on 5V power options. But since the parts come
with a two week delay, I'm looking at widening my options and 3.3V I/O on
data with 12V for blinkage seems to be taunting the Angry Blinky Ball in
ways I don't even yet imagine.

I've given more than a casual roam through AMZ/Ali and can't find a JST
connector in that pitch and size. (Turns out I have a zillion
crimpable/composable 2.0 JST 2.0's, but they top out at 16 pins. That box
has otherwise not been opened.) It's my hope that, in the grand spirit of
interoperability, this pinout and connector came from somewhere and I just
don't recognize it and/or am lacking the right words to describe it
sufficiently to insert money and receive premade cables. Maybe, for
example, this pinout came from a Beagleboard (it's not; they use the Pi 40
pin "standard") or something even more niche and there's a whole
ecosystem/subculture around it that I just don't know about.

There's just a lot of hardware options under the NightDriver flag and I'm
trying to gear up a build to help with the support issues for things like
today's (yesterday's?) "RGB? BGR? WTH?" that is probably easy if I could
plug it all up together.

So, yes, I'm very interested in guidance on whether it's best to try to
plug in (and if so, how) to Mesmerizer to help debug production-ish issues.
Pinouts, cables, Daughterboards (or hats or whatever theyy're called here)
sources would similarly be welcome.

If the guidance is "fugedabowtit" and move on to a home-brew board, info on
how to software-rendezvous on things like IR remote and Mic inputs - both
of which I'm guessing have a substantial investment in supporting - would
be awesome.

Thanx,
RJL

**Comment by tikurahul** [comment]
Thanks for sending me a board. I am now speed-running videos on the Channel to understand how the source code is structured. 
I am also having a bit of fun re-learning C++ 😃 not having written any in the last couple of years.

Will try writing a simple effect next, and then start looking into the source code for the server.

Regards,
--
Rahul

**Comment by davepl** [comment]
Get enlisted in the source code, make sure you can build it (NightDriverLED.com <http://nightdriverled.com/>) and if you’re interested in working on the React or CPP, email me with the title of BOARD REQUEST and your shipping address!

_ Dave

**Comment by DiannaLeason** [comment]
Hi Dave,Sadly...I'm pretty new to C++....I'm afraid I wont be of much help working on the REACT or CPP projects...Keep up the Great work, though! I really enjoy your Videos! They have  inspired me to get back into doing some programming, after setting it aside about 20 years ago...(yea...I'm 70 years old...)
You have Inspired many of us to get out of our comfort zone!...And that's a good thing! 
Cheers,Dianna Leason


    On Sunday, July 23, 2023 at 02:14:22 PM PDT, David W Plummer ***@***.***> wrote:  
 
 

Get enlisted in the source code, make sure you can build it (NightDriverLED.com <http://nightdriverled.com/>) and if you’re interested in working on the React or CPP, email me with the title of BOARD REQUEST and your shipping address!

_ Dave

**Comment by robertlipe** [comment]
Dianna, if you're not ready to wade into the nibbles and bits yourself, perhaps one of the boards listed on the front page will light you up.

m5stickc(or plus) might be out at Amazon, but still looks like a pretty nifty board. Ditto Heltec Wifi Kit V3. Both have the advantage of being readily installable via https://plummerssoftwarellc.github.io/NightDriverStrip/ so you may not even have to build the software on your own. You should be able to dig around in include/globals.h and figure out which pin are used for the WS2812 configuration pins.  You should be able to wire them up breadboard or wire up HUB75 pins with a little extra digging.

Let us know what you're looking for (are mic, remote, HUB75, etc. important?) and maybe we can recommend an existing retail board for you.

**Comment by DiannaLeason** [comment]
Hi,I have a couple m5stick’s arriving tomorrow….have used Dave’s web based installer to load code onto an esp32 dev module…and have got the ws2812 string running.I’m slowly finding my way around VScode….I’ll get there!Finally found Dave’s sample code examples for the YouTube video’s on GitHub. That will help a lot!Cheers,Dianna Leason
    On Sunday, July 23, 2023 at 09:17:59 PM PDT, Robert Lipe ***@***.***> wrote:  
 
 


Dianna, if you're not ready to wade into the nibbles and bits yourself, perhaps one of the boards listed on the front page will light you up.

m5stickc(or plus) might be out at Amazon, but still looks like a pretty nifty board. Ditto Heltec Wifi Kit V3. Both have the advantage of being readily installable via https://plummerssoftwarellc.github.io/NightDriverStrip/ so you may not even have to build the software on your own. You should be able to dig around in include/globals.h and figure out which pin are used for the WS2812 configuration pins. You should be able to wire them up breadboard or wire up HUB75 pins with a little extra digging.

Let us know what you're looking for (are mic, remote, HUB75, etc. important?) and maybe we can recommend an existing retail board for you.

**Comment by robertlipe** [comment]
Cool! Sounds like you're well on your way. Here I was trying to help you
get started and you're already starting your hoard. :-)

In no time, you'll have to leave the cars out of the garage because dev
boards, light strips, and power supplies will be taking over your life!

**Comment by DiannaLeason** [comment]
Robert,Can you give me a clue as to how Dave does his fireworks effect?Making Progress...C++ is still a little opaque to me, but I'm learning!TIADianna

    On Sunday, July 23, 2023 at 10:09:02 PM PDT, Robert Lipe ***@***.***> wrote:  
 
 

Cool! Sounds like you're well on your way. Here I was trying to help you
get started and you're already starting your hoard. :-)

In no time, you'll have to leave the cars out of the garage because dev
boards, light strips, and power supplies will be taking over your life!

**Comment by davepl** [comment]
There’s a “particle system” of classes I wrote, like FadingObject, MovingObject, and so on.  The fireworks effect creates a “burst” of them at a random start point with random velocities, and it looks fairly pleasing, but it’s pretty simple!

-= Dave

**Comment by robertlipe** [comment]
In case it helps, Dianna, I don't think there is anything in the code
currently called "firework" anything. This may be why your code searching
is failing/frustrating.

There is a big table that maps human names (those that show on the web
interface) to code entry points in effects.cpp
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/src/effects.cpp>.
It's not obvious from a casual code search, but sometimes the name comes in
ADD_EFFECT:

        ADD_EFFECT(EFFECT_MATRIX_WAVEFORM,          WaveformEffect,
"WaveIn", 8);

and sometimes, there's an additional layer where

        ADD_EFFECT(EFFECT_MATRIX_PONG_CLOCK,        PatternPongClock);
uses the name in the actual constructor
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/0ba548e0e938f363f43c783afc07baaf4ab290a4/include/effects/matrix/PatternPongClock.h#L98>.
So not everything is actually in effects.cpp. This is actually described
somewhat in the doc
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/0ba548e0e938f363f43c783afc07baaf4ab290a4/README.md?plain=1#L29>
for new effects, but I'll admit that I didn't recognize it until I saw it
right now.

Macros like this are pretty uncommon in C++ code bases and confusion
opportunities like this help show why.  Where everything else in C++ is
super strongly typed, in macros ("ADD_EFFECT") we get things like " Any
additional parameters are passed" so you really can't count on the Nth
paramater meaning the same thing everywhere. It can be disorienting. (Sorry
if I just called someone's baby ugly...)

Dave's hint tells us that the code for "fireworks" is probably here
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/include/effects/strip/particles.h>.
(Without a debugger, I wouldn't have made that leap, either...)  Dave's
Garage has a bunch of short episodes of walking through some of the
effects. I probably saw them long ago, before actually getting involved in
the code, so they probably imprinted on me, but I dont' know how well they
match the details of the code today.

FWIW, I haven't explored the strip drawing effects myself much yet, either.
Only last week did I really get a working USE_STRIP build going in my own
lab. I had been working more on the matrix group of effects.

If you're "just" looking for a simple effect to get your head around, you
might find particles.h a bit chewy with the 20-ish classes in a single
file. (In defense, NOT doing that in this code would mean duplicating a
bunch of long boilerplate which can be tough to machine-process later, so
that's just kind of necessary.)

If you're just looking for an effect to carve up to assist in learning, you
might find something like the meteor
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/0ba548e0e938f363f43c783afc07baaf4ab290a4/include/effects/strip/meteoreffect.h>
effect to be a bit more classical. It's much easier to recognize the setup
points (Init), the points where something actionable happens (Reverse() -
as a leraning exercise, change the colors when they reverse or something)
and the critical Draw() that's called dozens of times a second to actually,
well, draw. It's a bit easier to recognize that the effects dont' actually
touch hardware - they just manipulate the leds array at g()->leds[NUM_LEDS]
which is an array (sorta) of CRGBs. It's a bit easier to get your head
around for a first effect. There are other examples in incude/effects/strip
that are both easy and complicated; you'll recognize each when you visit
them.

When learning or exploring an unfamiliar code base, GitHub's web interface
is actually pretty handy (finally). You can click on a symbol and see what
calls it and what it calls and where it's defined vs. declared or such.  In
larger code bases, it can be really helpful.

To continue talks along this line, it's probably best to move to
'discussions' instead of inside the announcement for boards being ready.

Good luck on your journey!

**Comment by DiannaLeason** [comment]
Thank you so much....A lot to chew on here!L8RDianna

    On Friday, August 4, 2023 at 01:10:46 AM PDT, Robert Lipe ***@***.***> wrote:  
 
 

In case it helps, Dianna, I don't think there is anything in the code
currently called "firework" anything. This may be why your code searching
is failing/frustrating.

There is a big table that maps human names (those that show on the web
interface) to code entry points in effects.cpp
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/src/effects.cpp>.
It's not obvious from a casual code search, but sometimes the name comes in
ADD_EFFECT:

 ADD_EFFECT(EFFECT_MATRIX_WAVEFORM, WaveformEffect,
"WaveIn", 8);

and sometimes, there's an additional layer where

 ADD_EFFECT(EFFECT_MATRIX_PONG_CLOCK, PatternPongClock);
uses the name in the actual constructor
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/0ba548e0e938f363f43c783afc07baaf4ab290a4/include/effects/matrix/PatternPongClock.h#L98>.
So not everything is actually in effects.cpp. This is actually described
somewhat in the doc
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/0ba548e0e938f363f43c783afc07baaf4ab290a4/README.md?plain=1#L29>
for new effects, but I'll admit that I didn't recognize it until I saw it
right now.

Macros like this are pretty uncommon in C++ code bases and confusion
opportunities like this help show why. Where everything else in C++ is
super strongly typed, in macros ("ADD_EFFECT") we get things like " Any
additional parameters are passed" so you really can't count on the Nth
paramater meaning the same thing everywhere. It can be disorienting. (Sorry
if I just called someone's baby ugly...)

Dave's hint tells us that the code for "fireworks" is probably here
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/include/effects/strip/particles.h>.
(Without a debugger, I wouldn't have made that leap, either...) Dave's
Garage has a bunch of short episodes of walking through some of the
effects. I probably saw them long ago, before actually getting involved in
the code, so they probably imprinted on me, but I dont' know how well they
match the details of the code today.

FWIW, I haven't explored the strip drawing effects myself much yet, either.
Only last week did I really get a working USE_STRIP build going in my own
lab. I had been working more on the matrix group of effects.

If you're "just" looking for a simple effect to get your head around, you
might find particles.h a bit chewy with the 20-ish classes in a single
file. (In defense, NOT doing that in this code would mean duplicating a
bunch of long boilerplate which can be tough to machine-process later, so
that's just kind of necessary.)

If you're just looking for an effect to carve up to assist in learning, you
might find something like the meteor
<https://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/0ba548e0e938f363f43c783afc07baaf4ab290a4/include/effects/strip/meteoreffect.h>
effect to be a bit more classical. It's much easier to recognize the setup
points (Init), the points where something actionable happens (Reverse() -
as a leraning exercise, change the colors when they reverse or something)
and the critical Draw() that's called dozens of times a second to actually,
well, draw. It's a bit easier to recognize that the effects dont' actually
touch hardware - they just manipulate the leds array at g()->leds[NUM_LEDS]
which is an array (sorta) of CRGBs. It's a bit easier to get your head
around for a first effect. There are other examples in incude/effects/strip
that are both easy and complicated; you'll recognize each when you visit
them.

When learning or exploring an unfamiliar code base, GitHub's web interface
is actually pretty handy (finally). You can click on a symbol and see what
calls it and what it calls and where it's defined vs. declared or such. In
larger code bases, it can be really helpful.

To continue talks along this line, it's probably best to move to
'discussions' instead of inside the announcement for boards being ready.

Good luck on your journey!

**Comment by DiannaLeason** [comment]
Dave,Will you be putting any examples of this in the Github repository?TIADianna

    On Thursday, August 3, 2023 at 12:36:14 PM PDT, David W Plummer ***@***.***> wrote:  
 
 

There’s a “particle system” of classes I wrote, like FadingObject, MovingObject, and so on. The fireworks effect creates a “burst” of them at a random start point with random velocities, and it looks fairly pleasing, but it’s pretty simple!

-= Dave

**Comment by robertlipe** [comment]
I provided a link above.

**Comment by davepl** [comment]
Oh wait, you know what?  I’m travelling so working from memory, but maybe Fireworks only exists as a server-side effect.  I use it to send it over WiFi to my strips configured as LEDSTRIP projects.  That’s way more complicated than you want to get into, I’m sure.

But long story short, that means the code lives as C# code in NightDriverServer (already on Github).  So you could likely port it to C++ pretty easily, if you really wanted a good fireworks effect.

- Dave

**Comment by DiannaLeason** [comment]
Thanks Dave! That’s a lot to unpack! I’ve been using the web installer for my M5Sticks…never realized the source code was on GitHub!This will take a while to sort through…I’m intrigued by your use of the server to send the effects to the controllers in real time!Thanks and safe travels!Dianna
    On Friday, August 4, 2023 at 05:17:14 PM PDT, David W Plummer ***@***.***> wrote:  
 
 

Oh wait, you know what? I’m travelling so working from memory, but maybe Fireworks only exists as a server-side effect. I use it to send it over WiFi to my strips configured as LEDSTRIP projects. That’s way more complicated than you want to get into, I’m sure.

But long story short, that means the code lives as C# code in NightDriverServer (already on Github). So you could likely port it to C++ pretty easily, if you really wanted a good fireworks effect.

- Dave
